File: fraiseql-csharp/tests/FraiseQL.Security.Tests/AttributeBasedAccessControlTest.cs
using Xunit;
using FraiseQL.Security;
using System.Collections.Generic;

namespace FraiseQL.Security.Tests
{
    public class AttributeBasedAccessControlTest
    {
        [Fact]
        public void ShouldCreateABACPolicy()
        {
            var config = new AuthzPolicyBuilder("accessControl")
                .Type(new AuthzPolicyType.Abac())
                .Attributes("clearance_level >= 2")
                .Description("Basic clearance")
                .Build();

            Assert.Equal("accessControl", config.Name);
            Assert.IsType<AuthzPolicyType.Abac>(config.Type);
        }

        [Fact]
        public void ShouldHandleMultipleAttributes()
        {
            var config = new AuthzPolicyBuilder("secretAccess")
                .Type(new AuthzPolicyType.Abac())
                .Attributes("clearance_level >= 3", "background_check == true")
                .Build();

            Assert.Equal(2, config.Attributes.Count);
        }

        [Fact]
        public void ShouldCreateClearanceLevelPolicy()
        {
            var config = new AuthzPolicyBuilder("topSecret")
                .Type(new AuthzPolicyType.Abac())
                .Attributes("clearance_level >= 3")
                .Description("Top secret clearance required")
                .Build();

            Assert.Single(config.Attributes);
        }

        [Fact]
        public void ShouldCreateDepartmentPolicy()
        {
            var config = new AuthzPolicyBuilder("financeDept")
                .Type(new AuthzPolicyType.Abac())
                .Attributes("department == \"finance\"")
                .Description("Finance department only")
                .Build();

            Assert.Equal("financeDept", config.Name);
        }

        [Fact]
        public void ShouldCreateTimeBasedPolicy()
        {
            var config = new AuthzPolicyBuilder("businessHours")
                .Type(new AuthzPolicyType.Abac())
                .Attributes("now >= 9:00 AM", "now <= 5:00 PM")
                .Description("During business hours")
                .Build();

            Assert.Equal(2, config.Attributes.Count);
        }

        [Fact]
        public void ShouldCreateGeographicPolicy()
        {
            var config = new AuthzPolicyBuilder("usOnly")
                .Type(new AuthzPolicyType.Abac())
                .Attributes("country == \"US\"")
                .Description("United States only")
                .Build();

            Assert.Single(config.Attributes);
        }

        [Fact]
        public void ShouldCreateGDPRPolicy()
        {
            var config = new AuthzPolicyBuilder("gdprCompliance")
                .Type(new AuthzPolicyType.Abac())
                .Attributes("gdpr_compliant == true", "data_residency == \"EU\"")
                .Description("GDPR compliance required")
                .Build();

            Assert.Equal(2, config.Attributes.Count);
        }

        [Fact]
        public void ShouldCreateDataClassificationPolicy()
        {
            var config = new AuthzPolicyBuilder("classifiedData")
                .Type(new AuthzPolicyType.Abac())
                .Attributes("classification >= 2")
                .Description("For classified documents")
                .Build();

            Assert.Single(config.Attributes);
        }

        [Fact]
        public void ShouldSupportCachingInABACPolicy()
        {
            var config = new AuthzPolicyBuilder("cachedAccess")
                .Type(new AuthzPolicyType.Abac())
                .Attributes("role == \"viewer\"")
                .Cacheable(true)
                .CacheDurationSeconds(600)
                .Build();

            Assert.True(config.Cacheable);
            Assert.Equal(600, config.CacheDurationSeconds);
        }

        [Fact]
        public void ShouldSupportAuditLoggingInABACPolicy()
        {
            var config = new AuthzPolicyBuilder("auditedAccess")
                .Type(new AuthzPolicyType.Abac())
                .Attributes("audit_enabled == true")
                .AuditLogging(true)
                .Build();

            Assert.True(config.AuditLogging);
        }

        [Fact]
        public void ShouldSupportRecursiveApplicationInABACPolicy()
        {
            var config = new AuthzPolicyBuilder("recursiveAccess")
                .Type(new AuthzPolicyType.Abac())
                .Attributes("permission >= 1")
                .Recursive(true)
                .Description("Applies to nested types")
                .Build();

            Assert.True(config.Recursive);
        }

        [Fact]
        public void ShouldSetOperationSpecificAttributePolicy()
        {
            var config = new AuthzPolicyBuilder("readOnly")
                .Type(new AuthzPolicyType.Abac())
                .Attributes("can_read == true")
                .Operations("read")
                .Build();

            Assert.Equal("read", config.Operations);
        }

        [Fact]
        public void ShouldHandleAttributesArray()
        {
            var attrs = new List<string> { "attr1 >= 1", "attr2 == true" };
            var config = new AuthzPolicyBuilder("arrayTest")
                .Type(new AuthzPolicyType.Abac())
                .AttributesArray(attrs)
                .Build();

            Assert.Equal(2, config.Attributes.Count);
        }

        [Fact]
        public void ShouldCreateComplexABACPolicy()
        {
            var config = new AuthzPolicyBuilder("complex")
                .Type(new AuthzPolicyType.Abac())
                .Attributes("level >= 2", "verified == true", "active == true")
                .Description("Complex attribute rules")
                .AuditLogging(true)
                .Cacheable(true)
                .Build();

            Assert.Equal(3, config.Attributes.Count);
            Assert.True(config.AuditLogging);
        }

        [Fact]
        public void ShouldSetErrorMessageInABACPolicy()
        {
            var config = new AuthzPolicyBuilder("restricted")
                .Type(new AuthzPolicyType.Abac())
                .Attributes("clearance >= 2")
                .ErrorMessage("Insufficient clearance level")
                .Build();

            Assert.Equal("Insufficient clearance level", config.ErrorMessage);
        }
    }
}

File: fraiseql-php/tests/AttributeBasedAccessControlTest.php
<?php

declare(strict_types=1);

namespace FraiseQL\Tests;

use PHPUnit\Framework\TestCase;
use FraiseQL\SchemaRegistry;
use FraiseQL\Security\AuthzPolicyBuilder;
use FraiseQL\Security\AuthzPolicyType;
use FraiseQL\Attributes\AuthzPolicy;

/**
 * Tests for attribute-based authorization policies.
 */
final class AttributeBasedAccessControlTest extends TestCase
{
    protected function tearDown(): void
    {
        SchemaRegistry::getInstance()->clear();
        parent::tearDown();
    }

    public function testAbacPolicyDefinition(): void
    {
        $config = AuthzPolicyBuilder::create('secretClearance')
            ->type(AuthzPolicyType::ABAC)
            ->description('Requires top secret clearance')
            ->attributesArray(['clearance_level >= 3', 'background_check == true'])
            ->build();

        $this->assertSame('secretClearance', $config->name);
        $this->assertSame(AuthzPolicyType::ABAC, $config->type);
        $this->assertCount(2, $config->attributes);
    }

    public function testAbacAttributesVariadic(): void
    {
        $config = AuthzPolicyBuilder::create('financialData')
            ->attributes(
                'clearance_level >= 2',
                'department == "finance"',
                'mfa_enabled == true'
            )
            ->build();

        $this->assertCount(3, $config->attributes);
        $this->assertContains('clearance_level >= 2', $config->attributes);
    }

    public function testAbacAttributesArray(): void
    {
        $config = AuthzPolicyBuilder::create('regionalData')
            ->attributesArray(['region == "US"', 'gdpr_compliant == true'])
            ->build();

        $this->assertCount(2, $config->attributes);
    }

    public function testAbacClearanceLevelPattern(): void
    {
        $config = AuthzPolicyBuilder::create('classifiedDocument')
            ->type(AuthzPolicyType::ABAC)
            ->description('Access based on clearance level')
            ->attributes('clearance_level >= 2')
            ->build();

        $this->assertSame(AuthzPolicyType::ABAC, $config->type);
        $this->assertCount(1, $config->attributes);
    }

    public function testAbacDepartmentPattern(): void
    {
        $config = AuthzPolicyBuilder::create('departmentData')
            ->type(AuthzPolicyType::ABAC)
            ->attributes('department == "HR"')
            ->description('HR department access only')
            ->build();

        $this->assertSame('departmentData', $config->name);
    }

    public function testAbacTimeBasedPattern(): void
    {
        $config = AuthzPolicyBuilder::create('timeRestrictedData')
            ->type(AuthzPolicyType::ABAC)
            ->attributes(
                'current_time > "09:00"',
                'current_time < "17:00"',
                'day_of_week != "Sunday"'
            )
            ->description('Business hours access')
            ->build();

        $this->assertCount(3, $config->attributes);
    }

    public function testAbacGeographicPattern(): void
    {
        $config = AuthzPolicyBuilder::create('geographicRestriction')
            ->type(AuthzPolicyType::ABAC)
            ->attributes('region in ["US", "CA", "MX"]')
            ->description('North American access only')
            ->build();

        $this->assertCount(1, $config->attributes);
    }

    public function testAbacGdprCompliance(): void
    {
        $config = AuthzPolicyBuilder::create('personalData')
            ->type(AuthzPolicyType::ABAC)
            ->attributes(
                'gdpr_compliant == true',
                'data_residency == "EU"',
                'consent_given == true'
            )
            ->description('GDPR-compliant access')
            ->build();

        $this->assertCount(3, $config->attributes);
    }

    public function testAbacProjectBasedPattern(): void
    {
        $config = AuthzPolicyBuilder::create('projectData')
            ->type(AuthzPolicyType::ABAC)
            ->attributes('user_project == resource_project')
            ->description('Users can only access their own projects')
            ->build();

        $this->assertCount(1, $config->attributes);
    }

    public function testAbacClassificationPattern(): void
    {
        $config = AuthzPolicyBuilder::create('dataClassification')
            ->type(AuthzPolicyType::ABAC)
            ->attributes(
                'user_classification >= resource_classification',
                'has_need_to_know == true'
            )
            ->description('Classification-based access control')
            ->build();

        $this->assertCount(2, $config->attributes);
    }

    public function testAbacCaching(): void
    {
        $config = AuthzPolicyBuilder::create('cachedAbac')
            ->type(AuthzPolicyType::ABAC)
            ->attributes('attribute1 == "value"')
            ->cacheable(true)
            ->cacheDurationSeconds(3600)
            ->build();

        $this->assertTrue($config->cacheable);
        $this->assertSame(3600, $config->cacheDurationSeconds);
    }

    public function testAbacNoCache(): void
    {
        $config = AuthzPolicyBuilder::create('sensitivAbac')
            ->type(AuthzPolicyType::ABAC)
            ->attributes('sensitive_attribute == true')
            ->cacheable(false)
            ->build();

        $this->assertFalse($config->cacheable);
    }

    public function testAbacAuditLogging(): void
    {
        $config = AuthzPolicyBuilder::create('auditedAbac')
            ->type(AuthzPolicyType::ABAC)
            ->attributes('access_control == true')
            ->auditLogging(true)
            ->build();

        $this->assertTrue($config->auditLogging);
    }

    public function testAbacErrorMessage(): void
    {
        $config = AuthzPolicyBuilder::create('restrictedAbac')
            ->type(AuthzPolicyType::ABAC)
            ->attributes('clearance_level >= 3')
            ->errorMessage('Your clearance level is insufficient for this resource')
            ->build();

        $this->assertSame('Your clearance level is insufficient for this resource', $config->errorMessage);
    }

    public function testAbacOperationSpecific(): void
    {
        $config = AuthzPolicyBuilder::create('deleteRestricted')
            ->type(AuthzPolicyType::ABAC)
            ->attributes('role == "admin"')
            ->operations('delete,create')
            ->build();

        $this->assertSame('delete,create', $config->operations);
    }

    public function testAbacRecursive(): void
    {
        $config = AuthzPolicyBuilder::create('recursiveAbac')
            ->type(AuthzPolicyType::ABAC)
            ->attributes('hierarchy_level >= 2')
            ->recursive(true)
            ->build();

        $this->assertTrue($config->recursive);
    }

    public function testAbacFluentChaining(): void
    {
        $config = AuthzPolicyBuilder::create('complexAbac')
            ->type(AuthzPolicyType::ABAC)
            ->description('Complex ABAC policy')
            ->attributes('clearance >= 2', 'department == "IT"', 'mfa == true')
            ->cacheable(true)
            ->cacheDurationSeconds(1800)
            ->recursive(false)
            ->operations('read,update')
            ->auditLogging(true)
            ->errorMessage('Access denied')
            ->build();

        $this->assertSame('complexAbac', $config->name);
        $this->assertSame(AuthzPolicyType::ABAC, $config->type);
        $this->assertCount(3, $config->attributes);
        $this->assertTrue($config->cacheable);
        $this->assertTrue($config->auditLogging);
    }

    public function testAbacAttributeWithRule(): void
    {
        $config = AuthzPolicyBuilder::create('hybridAbac')
            ->type(AuthzPolicyType::ABAC)
            ->rule("hasAttribute(\$context, 'clearance_level', 3)")
            ->attributes('clearance_level >= 3')
            ->build();

        $this->assertSame("hasAttribute(\$context, 'clearance_level', 3)", $config->rule);
    }

    public function testAbacAttributePattern(): void
    {
        #[AuthzPolicy(
            name: 'abacExample',
            type: AuthzPolicyType::ABAC,
            attributes: ['clearance >= 2', 'department == "Finance"']
        )]
        class AbacExample
        {
        }

        $this->assertTrue(true);
    }
}
